introduce max_block_value_to_verify in constraints config

gd-0 opened the pull request:
- introduce new constraints config param: `max_block_value_to_verify`
- if the block value is greater than this field, inclusion constraints are not verified in `submit_block` and `get_header_with_proofs`

Reasoning behind this is that we don't want validators to miss out on big ETH rewards on mainnet while testing inclusion preconf blocks.

gd-0 committed:
introduce max_block_value_to_verify in constraints config

thedevbirb commented on crates/api/src/builder/api.rs:
```diff
@@ -2197,6 +2165,61 @@ where
+                info!(
+                    request_id = %request_id,
+                    block_value = %payload.value(),
+                    "block value is greater than max value to verify, skipping inclusion proof verification",
```
```suggestion
                    "block value is greater than max value to verify, inclusion proof verification and saving is skipped",
```

thedevbirb commented on crates/api/src/proposer/api.rs:
```diff
@@ -571,7 +571,21 @@ where
+                            %request_id,
+                            slot,
+                            value = ?bid.value(),
+                            "block value is greater than max value to verify, skipping inclusion proof check in get_header_with_proofs",
```
fyi inclusion proofs are not verified in the `get_header_with_proofs` call on the relay side.

gd-0 replied:
ah right yeah comment was a bit misleading here - have adjusted the comment

thedevbirb commented on crates/api/src/builder/api.rs:
```diff
@@ -2197,6 +2165,61 @@ where
+                );
+                return Ok(());
+            }
+        }
```
The function to verify and save inclusion proofs is great ty, but can we move this if check outside the function, before it is called? So that calling this functions means doing these steps regardless of the input.

thedevbirb commented on crates/api/src/proposer/api.rs:
```diff
@@ -571,7 +571,21 @@ where
+                            value = ?bid.value(),
+                            "block value is greater than max value to verify, skipping inclusion proof check in get_header_with_proofs",
+                        );
+                    }
```
I think this check can be done way earlier, so the flow can look like this:
- get the bid
- save trace to DB
- do the bid value check -- if it's satisfied, then do an early return since we the `submit_block_with_proofs` code doesn't save or verify proofs in that scenario
- if it's not satisfied, continue with the usual steps.

thedevbirb commented on crates/common/src/config.rs:
```diff
@@ -136,6 +136,10 @@ pub struct ConstraintsApiConfig {
+    /// We do this to ensure that high value blocks are not rejected.
+    #[serde(default)]
+    pub max_block_value_to_verify: Option<u64>,
 }
```
This change here results in the code not compiling because this field is missing in the implementation of the `Default` trait for `ConstraintsApiConfig`. Can you remove the `serde` attribute and add the default value `None` in the function below?

thedevbirb reviewed (changes requested):
Thanks for the PR! Here are some minor changes that should be done in our opinion.

Once you've done these, can you run `just fmt` and make sure `just fmt-check` and `just clippy` introduced in https://github.com/gattaca-com/helix/pull/48 run without errors?

gd-0 committed:
address PR comments + other fixes

- fix lints + other refactoring
- max_block_value_to_verify_wei now U256 to handle big value cases
- return earlier in get_header_with_proofs call

thedevbirb commented on crates/api/src/builder/api.rs:
```diff
@@ -679,42 +679,23 @@ where
-            .await?;
-
-            info!(request_id = %request_id, head_slot, "inclusion proofs verified and saved to auctioneer");
+        let skip_inclusion_proof_verify_and_save = api
+            .relay_config
+            .constraints_api_config
+            .max_block_value_to_verify_wei
+            .map_or(false, |max_block_value_to_verify| payload.value() > max_block_value_to_verify);
+        if !skip_inclusion_proof_verify_and_save {
+            if let Err(err) = api.verify_and_save_inclusion_proofs(&payload, &request_id).await {
+                warn!(request_id = %request_id, error = %err, "failed to verify and save inclusion proofs");
+                return Err(err)
+            }
```
What do you think of inverting the condition? Imo easier to read compared to a double negation
```rs
        let should_verify_and_save_inclusion_proofs = api
            .relay_config
            .constraints_api_config
            .max_block_value_to_verify_wei
            .map_or(true, |max_block_value_to_verify| payload.value() <= max_block_value_to_verify);
        if should_verify_and_save_inclusion_proofs {
            ...
         } else {
             info!("block value is greater than ...")
         }
```

thedevbirb commented on crates/api/src/test_utils.rs:
```diff
@@ -218,8 +218,8 @@ pub fn proposer_api_app() -> (
         Arc::new(ValidatorPreferences::default()),
-        0,
         gossip_receiver,
+        Arc::new(Default::default()),
```
This makes tests not compile, it should be
```suggestion
        Default::default(),
```

thedevbirb reviewed (changes requested):
Tested on the devnet and works good, a minor change is needed to make tests compile.

thedevbirb committed:
fix(tests): make them compile

thedevbirb committed:
chore(api): should verify proofs logic
